lisp/x/profiler/callgrind.go

```diff
@@ -183,67 +172,41 @@ func (p *callgrindProfiler) getCallRefAndDecrement() *callRef {
 	panic(fmt.Sprintf("Unset thread ref %d", *thread))
 }
 
-// Gets file parameters from the LVal
-func (p *callgrindProfiler) getFunctionParameters(function *lisp.LVal) (string, int, string) {
-	var source string
-	line := 0
-	if function.Source == nil {
-		if cell := function.Cells[0]; cell != nil && cell.Source != nil {
-			source = cell.Source.File
-			line = cell.Source.Line
-		} else {
-			source = "no-source"
-		}
-	} else {
-		source = function.Source.File
-		line = function.Source.Line
-	}
-	fName := fmt.Sprintf("%s:%s", function.FunData().Package, getFunNameFromFID(p.runtime, function.FunData().FID))
-	return source, line, fName
-}
-
-func (p *callgrindProfiler) end(function *lisp.LVal) {
+func (p *callgrindProfiler) end(fun *lisp.LVal) {
 	if !p.enabled {
 		return
 	}
 	p.Lock()
 	defer p.Unlock()
-	switch function.Type {
-	case lisp.LInt, lisp.LString, lisp.LFloat, lisp.LBytes, lisp.LError, lisp.LArray, lisp.LQuote, lisp.LNative, lisp.LSortMap:
-		// Again, we can ignore these as we never put them on the stack
-		return
-	case lisp.LFun, lisp.LSymbol, lisp.LSExpr, lisp.LQSymbol:
-		source, line, fName := p.getFunctionParameters(function)
-		// Write what function we've been observing and where to find it
-		fmt.Fprintf(p.writer, "fl=%s\n", p.getRef(source))
-		fmt.Fprintf(p.writer, "fn=%s\n", p.getRef(fName))
-		ref := p.getCallRefAndDecrement()
-		ref.duration = time.Since(ref.start)
-		if ref.duration == 0 {
-			ref.duration = 1
-		}
-		ms := &runtime.MemStats{}
-		runtime.ReadMemStats(ms)
-		ref.endMemory = ms.TotalAlloc
-		memory := ref.endMemory - ref.startMemory
-		// Cache the location - we won't be able to get it again when we build the maps for
-		// things that call this.
-		if ref.line == 0 && function.Source != nil {
-			ref.line = function.Source.Line
-			ref.file = function.Source.File
-		}
-		// Output timing and line ref
-		fmt.Fprintf(p.writer, "%d %d %d\n", line, ref.duration, memory)
-		// Output the things we called
-		for _, entry := range ref.children {
-			fmt.Fprintf(p.writer, "cfl=%s\n", p.getRef(entry.file))
-			fmt.Fprintf(p.writer, "cfn=%s\n", p.getRef(entry.name))
-			fmt.Fprint(p.writer, "calls=1 0 0\n")
-			fmt.Fprintf(p.writer, "%d %d %d\n", entry.line, entry.duration, memory)
-		}
-		// and end the entry
-		fmt.Fprint(p.writer, "\n")
-	default:
-		panic(fmt.Sprintf("Missing type %d", function.Type))
+	fName := prettyFunName(p.runtime, fun)
```

Comment:
Is this safe to call all the time? Looks like it can panic.

Reply:
Can it?I think I added a guard around it to avoid the panics--did I miss one?

Reply:
`prettyFunName` immediately calls `fun.FunData` which can panic depending on the type I believe

Reply:
Good call. Updated.